File: calculadora_cientifica.py
import os
import math
def suma(x, y):
  return x + y
def resta(x, y):
  return x - y
def multiplicar(x, y):
  return x * y
def dividir(x, y):
  if y == 0.0:
    return "Error"
  else:
    return x / y
def potencia(x, y):
  return x ** y
def raiz_cuadrada(x):
  return x ** 0.5
def factorial(x):
  if x == 0:
    return 1
  else:
    return x * factorial(x - 1)
def logaritmo_10(x):
  return math.log10(x)
def logaritmo(x, y):
  return math.log(y, x)
def seno(x):
  return math.sin(x)
def coseno(x):
  return math.cos(x)
def tangente(x):
  return math.tan(x)
def main():
  while True:
    os.system('cls')
    print("+ -> Suma")
    print("- -> Resta")
    print("* -> Multiplicar")
    print("/ -> Dividir")
    print("^ -> Potencia")
    print("sqrt -> Raiz cuadrada")
    print("! -> Factorial")
    print("log -> Logaritmo(10)")
    print("logx -> Logaritmo(x)")
    print("sin -> Seno")
    print("cos -> Coseno")
    print("tan -> Tangente")
    print(". -> Salir")
    choice = input("Escoje una opcion: ")
    if choice == "+":
      x = float(input("Introduce el primer numero: "))
      y = float(input("Introduce el segundo numero: "))
      print(x, "+", y, "=", suma(x, y))
    elif choice == "-":
      x = float(input("Introduce el primer numero: "))
      y = float(input("Introduce el segundo numero: "))
      print(x, "-", y, "=", resta(x, y))
    elif choice == "*":
      x = float(input("Introduce el primer numero: "))
      y = float(input("Introduce el segundo numero: "))
      print(x, "*", y, "=", multiplicar(x, y))
    elif choice == "/":
      x = float(input("Introduce el primer numero: "))
      y = float(input("Introduce el segundo numero: "))
      print(x, "/", y, "=", dividir(x, y))
    elif choice == "^":
      x = float(input("Introduce la base: "))
      y = float(input("Introduce el exponente: "))
      print(x, "^", y, "=", potencia(x, y))
    elif choice == "sqrt":
      x = float(input("Introduce el numero: "))
      print("√(",x,") =", raiz_cuadrada(x))
    elif choice == "!":
      x = int(input("Introduce el numero: "))
      print(x,"!", factorial(x))
    elif choice == "log":
      x = int(input("Introduce el numero: "))
      print("log(",x,") =", logaritmo_10(x))
    elif choice == "logx":
      x = int(input("Introduce la base del logaritmo: "))
      y = int(input("Introduce el numero: "))
      print("log(",y,") de base", x, "=", logaritmo(x, y))
    elif choice == "sin":
      x = int(input("Introduce el numero: "))
      print("sin(",x,") =", seno(x))
    elif choice == "cos":
      x = int(input("Introduce el numero: "))
      print("cos(",x,") =", coseno(x))
    elif choice == "tan":
      x = int(input("Introduce el numero: "))
      print("tan(",x,") =", tangente(x))
    elif choice == ".":
      break
    else:
      print("Error. Opcion incorrecta")
    continuar = input("Pulsa Enter para continuar: ")
if __name__ == "__main__":
  main()

File: Calculadora_cientifica.java
import java.util.regex.Pattern;
import javax.swing.JOptionPane;

public class Calculadora_cientifica {
	public static void main(String[] args) {
		String[] signos = {"+", "-", "*", "/", "^", "%", "√", "log10", "!", "sin", "cos", "tan", "salir"};
		
		String textnum2;
		double op1;
		String textnum= JOptionPane.showInputDialog("Introduce el primer operador");
		int signo = -1;

		if (Pattern.matches("[a-zA-Z]+", textnum) || textnum == null || textnum.length() == 0) {op1 = 0d;} else {op1 = Double.parseDouble(textnum);}
		
		while (signo != 12) {
			
			double op2 = 0;
			double result = 1;
			signo = JOptionPane.showOptionDialog(null, "¿Que quieres hacer?", null, 0, 3, null, signos, null);
			
			switch (signo) {
			case 0:
				textnum2 = JOptionPane.showInputDialog("Introduce el segundo operador");
				if (Pattern.matches("[a-zA-Z]+", textnum2) || textnum2 == null || textnum2.length() == 0) {op2 = 0d;} else {op2 = Double.parseDouble(textnum2);}
				JOptionPane.showMessageDialog(null, op1+" + "+op2+" = "+(op1+op2));
				op1 = op1+op2;
				continue;
			case 1:
				textnum2 = JOptionPane.showInputDialog("Introduce el segundo operador");
				if (Pattern.matches("[a-zA-Z]+", textnum2) || textnum2 == null || textnum2.length() == 0) {op2 = 0d;} else {op2 = Double.parseDouble(textnum2);}
				JOptionPane.showMessageDialog(null, op1+" - "+op2+" = "+(op1-op2));
				op1 = op1-op2;
				continue;
			case 2:
				textnum2 = JOptionPane.showInputDialog("Introduce el segundo operador");
				if (Pattern.matches("[a-zA-Z]+", textnum2) || textnum2 == null || textnum2.length() == 0) {op2 = 0d;} else {op2 = Double.parseDouble(textnum2);}
				JOptionPane.showMessageDialog(null, op1+" * "+op2+" = "+(op1*op2));
				op1 = op1*op2;
				continue;
			case 3:
				textnum2 = JOptionPane.showInputDialog("Introduce el segundo operador");
				if (Pattern.matches("[a-zA-Z]+", textnum2) || textnum2 == null || textnum2.length() == 0) {op2 = 0d;} else {op2 = Double.parseDouble(textnum2);}
				if (op2 == 0) {
					JOptionPane.showMessageDialog(null, "No se puede dividir entre 0","Error", 0);
				}else {
					JOptionPane.showMessageDialog(null, op1+" / "+op2+" = "+(op1/op2));
					op1 = op1/op2;
				}
				continue;
			case 4:
				textnum2 = JOptionPane.showInputDialog("Introduce el exponente");
				if (Pattern.matches("[a-zA-Z]+", textnum2) || textnum2 == null || textnum2.length() == 0) {op2 = 0d;} else {op2 = Double.parseDouble(textnum2);}
				JOptionPane.showMessageDialog(null, op1+" ^ "+op2+" = "+Math.pow(op1, op2));
				op1 = Math.pow(op1,op2);
				continue;
			case 5:
				textnum2 = JOptionPane.showInputDialog("Introduce el segundo operador");
				if (Pattern.matches("[a-zA-Z]+", textnum2) || textnum2 == null || textnum2.length() == 0) {op2 = 0d;} else {op2 = Double.parseDouble(textnum2);}
				if (op2 == 0) {
					JOptionPane.showMessageDialog(null, "No se puede hacer el resto entre 0","Error", 0);
				}else {
					JOptionPane.showMessageDialog(null, op1+" % "+op2+" = "+(op1%op2));
					op1 = op1%op2;
				}
				continue;
			case 6:
				if (op1 < 0) {
					JOptionPane.showMessageDialog(null, "No se puede hacer la raíz cuadrada de un numero negativo","Error", 0);
				}else {
					JOptionPane.showMessageDialog(null, "√("+op1+") = "+Math.sqrt(op1));
					op1 = Math.sqrt(op1);
				}
				continue;
			case 7:
				if (op1 <= 0) {
					JOptionPane.showMessageDialog(null, "No se puede hacer el logaritmo de base 10 de un numero negativo o 0","Error", 0);
				}else {
					JOptionPane.showMessageDialog(null, "log("+op1+") = "+Math.log10(op1));
					op1 = Math.log10(op1);
				}
				continue;
			case 8:
				if (op1 == 0) {
					JOptionPane.showMessageDialog(null, op1+"! = 1");
					op1 = 1d;					
				}else if (op1 < 0) {
					JOptionPane.showMessageDialog(null, "No se puede hacer el factorial de un numero negativo","Error", 0);
				}else {
					int num = (int) op1;
					for (int i = 2; i <= num; i++) {
						result *= i;			
					}
					JOptionPane.showMessageDialog(null, num+"! = "+result);
					op1 = result;
				}
				continue;
			case 9:
				JOptionPane.showMessageDialog(null, "sin("+op1+") = "+Math.sin(op1));
				op1 = Math.sin(op1);
				continue;
			case 10:
				JOptionPane.showMessageDialog(null, "cos("+op1+") = "+Math.cos(op1));
				op1 = Math.cos(op1);
				continue;
			case 11:
				JOptionPane.showMessageDialog(null, "tan("+op1+") = "+Math.tan(op1));
				op1 = Math.tan(op1);
				continue;
			case 12:
				System.out.println("Has salido");
				continue;
			default:
				JOptionPane.showMessageDialog(null, "Signo erroneo", null, 0);
				continue;
			}
		}
		
		
	}
}

